doc/sphinx/Pacemaker_Development/c.rst

```diff
@@ -15,6 +15,203 @@ Some existing Pacemaker code does not follow these guidelines, for historical
+| libstonithd   | stonith | lib/fencing   | | include/crm/stonith-ng.h| .. index::                          |
+|               |         |               | | include/crm/fencing/*   |    single: C library; libstonithd   |
+|               |         |               |                           |    single: libstonithd              |
+|               |         |               |                           |                                     |
+|               |         |               |                           | API for pacemaker-fenced IPC        |
++---------------+---------+---------------+---------------------------+-------------------------------------+
+
+
+Public versus Internal APIs
+___________________________
+
+Pacemaker libraries have both internal and public APIs. Internal APIs are those
+used only within Pacemaker; public APIs are those offered (via header files and
+documentation) for external code to use.
+
+Generic functionality needed by Pacemaker itself, such as string processing or
+XML processing, should remain internal, while functions providing useful
+high-level access to Pacemaker capabilities should be public. When in doubt,
+keep APIs internal, because it's easier to expose a previously internal API
+than hide a previously public API.
+
+Internal APIs can be changed as needed.
+
+The public API/ABI should maintain a degree of stability so that external
+applications using it do not need to be rewritten or rebuilt frequently. Many
+OSes/distributions avoid breaking API/ABI compatibility within a major release,
+so if Pacemaker breaks compatibility, that significantly delays when OSes
+can package the new version. Therefore, changes to public APIs should be
+backward-compatible (as detailed throughout this chapter), unless we are doing
+a (rare) release where we specifically intend to break compatibility.
+
+External applications known to use Pacemaker's public C API include
+`sbd <https://github.com/ClusterLabs/sbd>`_ and dlm_controld.
+
+
+.. index::
+   pair: C; API documentation
+   single: Doxygen
+
+API Documentation
+_________________
+
+Pacemaker uses `Doxygen <https://www.doxygen.nl/manual/docblocks.html>`_
+to automatically generate its
+`online API documentation <https://clusterlabs.org/pacemaker/doxygen/>`_,
+so all public API (header files, functions, structs, enums, etc.) should be
+documented with Doxygen comment blocks. Other code may be documented in the
+same way if desired, with an ``\internal`` tag in the Doxygen comment.
+
+Simple example of an internal function with a Doxygen comment block:
+
+.. code-block:: c
+
+   /*!
+    * \internal
+    * \brief Return string length plus 1
+    *
+    * Return the number of characters in a given string, plus one.
+    *
+    * \param[in] s  A string (must not be NULL)
+    *
+    * \return The length of \p s plus 1.
+    */
+   static int
+   f(const char *s)
```

Comment:
We don't do this currently, but I'd love to see us start and put it in the docs so new people start doing it too... this function could be marked with `__attribute__((nonnull))`.

Reply:
Hmm, maybe. There's no G_GNUC_* for it so we'd have to define something similar in configure.ac. It would help catch some compile-time issues but we'd still need run-time checks in the code, so I'm not sure it's worth the clutter.